File: problem/Valid_Parentheses/isValid.js
var isValid = function(s) {
    const stack = [];
    const mapping = {
        ')': '(',
        '}': '{',
        ']': '['
    };
    
    for (let char of s) {
        if (char in mapping) {  
            if (stack.length === 0 || stack.pop() !== mapping[char]) {
                return false;
            }
        } else {  
            stack.push(char);
        }
    }
    
    return stack.length === 0;
};

var isValid_optimized = function(s) {
    const stack = [];
    
    for (let char of s) {
        if (char === '(' || char === '[' || char === '{') {
            stack.push(char);
        } else {
            if (stack.length === 0) return false;
            
            const top = stack.pop();
            
            if ((char === ')' && top !== '(') ||
                (char === ']' && top !== '[') ||
                (char === '}' && top !== '{')) {
                return false;
            }
        }
    }
    
    return stack.length === 0;
};

var isValid_replacement = function(s) {
    while (s.includes('()') || s.includes('[]') || s.includes('{}')) {
        s = s.replace('()', '').replace('[]', '').replace('{}', '');
    }
    return s === '';
};

File: problem/Valid_Parentheses/isValid.cpp
class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> st;
        unordered_map<char, char> mapping = {
            {')', '('}, {'}', '{'}, {']', '['}};

        for (char c : s)
        {
            if (mapping.find(c) != mapping.end())
            {
                if (st.empty() || st.top() != mapping[c])
                {
                    return false;
                }
                st.pop();
            }
            else
            {
                st.push(c);
            }
        }

        return st.empty();
    }

    bool isValid_optimized(string s)
    {
        stack<char> st;

        for (char c : s)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                st.push(c);
            }
            else
            {
                if (st.empty())
                    return false;

                char top = st.top();
                st.pop();

                if ((c == ')' && top != '(') ||
                    (c == ']' && top != '[') ||
                    (c == '}' && top != '{'))
                {
                    return false;
                }
            }
        }

        return st.empty();
    }
};
